presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetCompressionUtils.java

```diff
@@ -48,35 +54,70 @@ public static Slice decompress(CompressionCodecName codec, Slice input, int unco
                 return decompressSnappy(input, uncompressedSize);
             case UNCOMPRESSED:
                 return input;
+            case LZO:
+                return decompressLZO(input, uncompressedSize);
             default:
-                // TODO add LZO support using pure java aircompressor
                 throw new ParquetCorruptionException("Codec not supported in Parquet: " + codec);
         }
     }
 
     private static Slice decompressSnappy(Slice input, int uncompressedSize)
     {
-        byte[] inArray = (byte[]) input.getBase();
-        int inOffset = (int) (input.getAddress() - ARRAY_BYTE_BASE_OFFSET);
-        int inLength = input.length();
-
         byte[] buffer = new byte[uncompressedSize];
-        SnappyDecompressor decompressor = new SnappyDecompressor();
-        decompressor.decompress(inArray, inOffset, inLength, buffer, 0, uncompressedSize);
+        decompress(new SnappyDecompressor(), input, 0, input.length(), buffer, 0);
         return wrappedBuffer(buffer);
     }
 
     private static Slice decompressGzip(Slice input, int uncompressedSize)
             throws IOException
     {
-        ByteArrayOutputStream uncompressedStream = new ByteArrayOutputStream();
+        DynamicSliceOutput sliceOutput = new DynamicSliceOutput(uncompressedSize);
+
+        if (uncompressedSize == 0) {
+            return sliceOutput.getUnderlyingSlice();
+        }
+
         byte[] buffer = new byte[uncompressedSize];
-        try (InputStream gzipInputStream = new GZIPInputStream(new ByteArrayInputStream(input.getBytes()))) {
+        try (InputStream gzipInputStream = new GZIPInputStream(input.getInput(), GZIP_BUFFER_SIZE)) {
             int bytesRead;
             while ((bytesRead = gzipInputStream.read(buffer)) != -1) {
-                uncompressedStream.write(buffer, 0, bytesRead);
+                sliceOutput.write(buffer, 0, bytesRead);
+            }
+            return sliceOutput.getUnderlyingSlice();
+        }
+    }
+
+    private static Slice decompressLZO(Slice input, int uncompressedSize)
+    {
+        LzoDecompressor lzoDecompressor = new LzoDecompressor();
+        long totalDecompressedCount = 0;
+        // over allocate buffer which makes decompression easier
+        byte[] output = new byte[uncompressedSize + SIZE_OF_LONG];
+        int outputOffset = 0;
+        int inputOffset = 0;
+        int cumulativeUncompressedBlockLength = 0;
+
+        while (totalDecompressedCount < uncompressedSize) {
+            if (totalDecompressedCount == cumulativeUncompressedBlockLength) {
```

Comment:
@dain good catch regarding the check. If we keep track of the `cumulativeUncompressedBlockLength` the check should become correct.